crates/extensions/c8y_mapper_ext/src/operations/handlers/custom_operation.rs

```diff
@@ -15,45 +20,110 @@ impl OperationContext {
+        let mapper_id = self.command_id.prefix();
+        let operation = match self
+            .get_c8y_operation(mapper_id, &command.payload)
+            .context("Could not get CumulocitySupportedOperation from payload")
+        {
+            Ok(c8y_operation) => c8y_operation,
+            Err(_) => return (Ok(OperationOutcome::Ignored), None),
+        };
 
-        match command.get_command_status() {
+        let sm_topic = &target.smartrest_publish_topic;
+        let result = match command.get_command_status() {
             CommandStatus::Executing => Ok(OperationOutcome::Executing {
                 extra_messages: vec![],
             }),
             CommandStatus::Successful => {
-                let smartrest_set_operation = self.get_smartrest_successful_status_payload(
-                    CumulocitySupportedOperations::C8yCustom(operation_name.to_string()),
-                    cmd_id,
-                );
+                let smartrest_set_operation =
+                    self.convert_output(mapper_id, &command, operation.clone(), cmd_id);
                 let c8y_notification = MqttMessage::new(sm_topic, smartrest_set_operation);
 
                 Ok(OperationOutcome::Finished {
                     messages: vec![c8y_notification],
                 })
             }
             CommandStatus::Failed { reason } => {
-                let smartrest_set_operation = self.get_smartrest_failed_status_payload(
-                    CumulocitySupportedOperations::C8yCustom(operation_name.to_string()),
-                    &reason,
-                    cmd_id,
-                );
+                let smartrest_set_operation =
+                    self.get_smartrest_failed_status_payload(operation.clone(), &reason, cmd_id);
                 let c8y_notification = MqttMessage::new(sm_topic, smartrest_set_operation);
 
                 Ok(OperationOutcome::Finished {
                     messages: vec![c8y_notification],
                 })
             }
             _ => Ok(OperationOutcome::Ignored),
+        };
+
+        (result, Some(operation))
+    }
+
+    fn get_c8y_operation(
+        &self,
+        mapper_id: &str,
+        value: &serde_json::Value,
+    ) -> Option<CumulocitySupportedOperations> {
+        if let Some(value) = value.get(mapper_id) {
+            if let Some(c8y_operation) = value.get("on_fragment") {
+                if let Some(c8y_operation) = c8y_operation.as_str() {
+                    return Some(CumulocitySupportedOperations::C8yCustom(
+                        c8y_operation.to_string(),
+                    ));
+                }
+            }
+        }
+        None
+    }
+
+    fn convert_output(
+        &self,
+        mapper_id: &str,
+        state: &GenericCommandState,
+        operation: CumulocitySupportedOperations,
+        cmd_id: &str,
+    ) -> SmartrestPayload {
+        match state.payload.get(mapper_id).unwrap().get("output") {
```

Comment:
The state being received from MQTT, it's not safe to assume the payload has a `c8y` fragment.

This has to be replaced by a function similar to `get_c8y_operation`.

Reply:
It's *currently* safe but only thanks to the *current* sequence of function calls, something that might change in the future.